crates/ide-assists/src/handlers/extract_function.rs

```diff
@@ -1264,6 +1276,34 @@ fn node_to_insert_after(body: &FunctionBody, anchor: Anchor) -> Option<SyntaxNod
     last_ancestor
 }
 
+fn find_non_trait_impl(trait_impl: &SyntaxNode) -> Option<SyntaxNode> {
+    let impl_type = Some(impl_type_name(trait_impl)?);
+
+    let mut sibblings = trait_impl.parent()?.children();
+    sibblings.find(|s| impl_type_name(s) == impl_type && !is_trait_impl(s))
+}
+
+fn last_impl_member(impl_node: &SyntaxNode) -> Option<SyntaxNode> {
+    impl_node.children().find(|c| c.kind() == SyntaxKind::ASSOC_ITEM_LIST)?.last_child()
+}
+
+fn is_trait_impl(node: &SyntaxNode) -> bool {
+    if !ast::Impl::can_cast(node.kind()) {
+        return false;
+    }
+    match ast::Impl::cast(node.clone()) {
+        Some(c) => c.trait_().is_some(),
+        None => false,
+    }
+}
+
+fn impl_type_name(impl_node: &SyntaxNode) -> Option<String> {
+    if !ast::Impl::can_cast(impl_node.kind()) {
+        return None;
+    }
+    Some(ast::Impl::cast(impl_node.clone())?.self_ty()?.to_string())
+}
```

Comment:
All of these seem to assume that the node is an `ast::Impl`, so I think we should just cast the node to one in `find_non_trait_impl`?

Reply:
Yep, you're right, that cleans things up a bit!